### DISCUSSION TITLE
How to compile OpenFAST for Simulink

### DISCUSSION THREAD

**Discussion by liruiqi97** [OpenFAST/openfast#1036] [Q&A] [answered]
I am trying to use OpenFAST on Simulink on my MacBook. When I was trying to build up the s-function using `create_FAST_SFunc`, the heading comment in this file tells that "we must have compiled OpenFAST for Simulink to create a DLL".

I have installed OpenFAST and all dependencies via the terminal, and I am trying to run the glue code (e.g. `Test01_SIG.m`). but I am not sure whether I have complied OpenFAST for Simulink and where the DLL (`.so`, `.dylib`, `.lib` etc) locates. How do we check that?

Thanks for any comment.

**Comment by bjonkman** [comment, marked as answer]
Do you have a library file called `openfastlib` in the `<installDir>/lib` directory (`installDir` being the variable specified in `create_FAST_SFunc.m`)?

**Reply by liruiqi97** [reply]
No, so I think I am essentially trying to build an `openfastlib` file. How can we build this file?

**Reply by bjonkman** [reply]
When you build OpenFAST, make sure you use `-DBUILD_OPENFAST_SIMULINK_API=ON`. See https://openfast.readthedocs.io/en/dev/source/install/index.html#openfast-cmake-options. This will create the library instead of an executable.

**Reply by liruiqi97** [reply]
all right... I eventually got those files by running the codes in [Documentation 2.2.5](https://raf-openfast.readthedocs.io/en/docs-fast8readme/source/install/index.html#cmake-with-make-for-linux-macos) to finish installation. I will test if it runs properly. Thanks for answering my questions.

**Reply by liruiqi97** [reply]
@bjonkman Hi Bonnie, would you mind to answer some extra questions? I just succeed building up the S-Function file `FAST_SFunc.mexmaci64` in the folder `openfast/install/bin`. Everything seems good, until an error is reported when I tried running the test `Run_Test01_SIG.m`.

The error code:
`Error using Run_Test01_SIG (line 36)
Error due to multiple causes.
Caused by:
    Error using Run_Test01_SIG (line 36)
    Error while obtaining sizes from MEX S-function 'FAST_SFunc' in 'Test01_SIG/FAST
    Nonlinear Wind Turbine/S-Function'.
        Error using Run_Test01_SIG (line 36)
        Invalid MEX-file '/Users/liruiqi/OpenFAST/install/lib/FAST_SFunc.mexmaci64':
        dlopen(/Users/liruiqi/OpenFAST/install/lib/FAST_SFunc.mexmaci64, 0x0006): Library
        not loaded: @rpath/libopenfastlib.dylib
          Referenced from: /Users/liruiqi/OpenFAST/install/lib/FAST_SFunc.mexmaci64
          Reason: tried:
          '/Applications/MATLAB_R2021b.app/bin/maci64/libopenfastlib.dylib' (no such
          file), '/Applications/MATLAB_R2021b.app/bin/maci64/./libopenfastlib.dylib' (no
          such file),
          '/Applications/MATLAB_R2021b.app/bin/maci64/../../sys/os/maci64/libopenfastlib.dylib'
          (no such file),
          '/Applications/MATLAB_R2021b.app/Contents/MacOS/libopenfastlib.dylib' (no such
          file), '/Applications/MATLAB_R2021b.app/Contents/MacOS/./libopenfastlib.dylib'
          (no such file),
          '/Applications/MATLAB_R2021b.app/Contents/MacOS/../../standalone/bin/maci64/libopenfastlib.dylib'
          (no such file),
          '/Applications/MATLAB_R2021b.app/Contents/MacOS/../../sys/os/maci64/libopenfastlib.dylib'
          (no such file),
          '/Applications/MATLAB_R2021b.app/Contents/MacOS/../../bin/maci64/libopenfastlib.dylib'
          (no such file), '/usr/local/lib/libopenfastlib.dylib' (no such file),
          '/usr/lib/libopenfastlib.dylib' (no such file)
    Error using Run_Test01_SIG (line 36)
    The expression: u(strmatch('LSSGagVxa',OutList))
    in 'Test01_SIG/Fcn'
    has a syntax error`

It seems that the MATLAB is looking for `libopenfastlib.dylib`, which file has been generated but not in the directories MATLAB tried to find it.

**Reply by liruiqi97** [reply]
@bjonkman Another issue I an encounting is that, when I tried to set `cmake .. -DBUILD_OPENFAST_SIMULINK_API=ON`, the terminal returns me some error. Error message:

`CMake Error at /usr/local/Cellar/cmake/3.22.3/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Matlab (missing: Matlab_INCLUDE_DIRS Matlab_MEX_LIBRARY
  Matlab_MEX_EXTENSION) (found version "unknown")
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.22.3/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindMatlab.cmake:1693 (find_package_handle_standard_args)
  modules/nwtc-library/CMakeLists.txt:127 (find_package)`

**Reply by bjonkman** [reply]
@liruiqi97, typically on Windows, I just make sure that the library is on the Matlab path (either by modifying the path or moving the file). I think macOS has made this more difficult recently, though. @andrew-platt, do you have any advice on this?

As for the error about not finding Matlab, the OpenFAST cmake files probably need to be updated for your version of Matlab. If you add the two highlighted lines below to the `cmake/FindMatlab.cmake` file in the OpenFAST repo, this will likely fix the problem:
![image](https://user-images.githubusercontent.com/6108781/157489085-d775087f-966d-400a-9179-4dbb4f86db0b.png)

**Reply by liruiqi97** [reply]
I tried copying (or moving) the library files (those in `OpenFAST/install/lib`) into one of the paths where MATLAB looks for the library. It does not report any error, BUT MATLAB eventually quits after busy loading for a while. (I guess this might be a bug on MacOS? or directly moving those files might not be a proper way to solve the problem on MacOS?)

**Reply by liruiqi97** [reply]
As for the second problem, I added those two lines in `cmake/FindMatlab.cmake` as Bonnie advises, but the same error still happens. The error code (same to my previous reply) in Terminal says that `Matlab_INCLUDE_DIRS`, `Matlab_MEX_LIBRARY`, `Matlab_MEX_EXTENSION` are all missing. When I check `OpenFAST/build/CMakeCache.txt`, those variables are all empty or unset:
<img width="436" alt="Screen Shot 2022-03-09 at 2 24 48 PM" src="https://user-images.githubusercontent.com/43281955/157516668-c17dfab5-469f-4d98-9bed-b8966d852105.png">
I guess I need to set those variables manually, but no idea which values should they be.

**Reply by bjonkman** [reply]
You could also try getting the updated FindMatlab.cmake file here: https://github.com/Kitware/CMake/blob/master/Modules/FindMatlab.cmake. Also, there is documentation here: https://cmake.org/cmake/help/latest/module/FindMatlab.html.

What version of Matlab are you using?

**Reply by liruiqi97** [reply]
still encounter the same error... thanks for sharing anyway! I am using MATLAB R2021b with a student license.

**Reply by bjonkman** [reply]
You could set those variables manually. `Matlab_ROOT_Dir` should be set to the value of what the `matlabroot` command returns in MATLAB.

**Reply by liruiqi97** [reply]
Okay I would have a try. Thanks Bonnie for all your help!

**Comment by Melancho7ee** [comment]
@liruiqi97 Sorry to bother, but I am wondering if you have solved this problem? I encoutered the same question these days. I dont know how to generate the "mexmaci64" file for OpenFAST-Simulink simulation, could you please help me?

**Reply by liruiqi97** [reply]
I have solved it and have run it successfully on Simulink on my MacBook. I am glad for giving any help.

**Comment by Melancho7ee** [comment]
Thanks for your reply, I am wondering how to get the mexmaci64 file. When I try to compile the openfast on my mac, some strange errors occurred (one of them is in the attachment). My mac is based on M1.&nbsp;
Could you please explain your methods? And by the way, are you Chinese? If so, maybe we could communicate in Chinese. Many Thanks for your quick reply. :-)


李佳奇
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "OpenFAST/openfast"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年10月31日(星期一) 晚上9:38
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [OpenFAST/openfast] How to compile OpenFAST for Simulink (Discussion #1036)





 
I have solved it and have run it successfully on Simulink on my MacBook. I am glad for giving any help.

**Reply by liruiqi97** [reply]
I was using cmake to install openfast on my computer, and then used `glue-code/simulink/src/create_FAST_SFunc.m` to create the `FAST_SFunc.mexmaci64` file. Not sure if you have finished the installation and found the location where the files were generated.
